### DISCUSSION TITLE
Conditional validation based on property

### DISCUSSION THREAD

**Discussion by chrisboakes** [colinhacks/zod#2207] [Q&A] [answered]
Hello,
I'm trying to validate a property based on another and I wondered the best way to achieve this.
Any help would be greatly appreciated 🙏 !

For example:

```js
const fooBar = z.object({
  foo: z.boolean(),
  // How do I validate this only if foo is true?
  bar: z.string().min(10, { message: "Should be 10 characters" }),
});

const schema = z.object({
  fields: z.array(fooBar),
});

test("should not validate bar when foo is false", (t) => {
  t.true(
    schema.safeParse({
      fields: [
        { foo: true, bar: "this is a valid field" },
        { foo: false, bar: "short" },
      ],
    }).success
  );
  t.end();
});
```

I've created a sandbox [here](https://codesandbox.io/p/sandbox/staging-silence-ojufdb?welcome=true)

**Comment by alexbrokethis** [comment]
You should be able to do that by making `bar` optional and then chaining a `refine` method to your `fooBar` schema:
```ts
const fooBar = z.object({
  foo: z.boolean(),
  bar: z.string().min(10, { message: "Should be 10 characters" }).optional(),
}).refine((value) => {
  if (value.foo === true) return !!value.bar
});
```

**Reply by chrisboakes** [reply]
Thanks @alexbrokethis. I tried updating my example [here](https://codesandbox.io/p/sandbox/staging-silence-ojufdb?welcome=true) but it still fails this test:

```js
test("should not validate bar when foo is false", (t) => {
  t.true(
    schema.safeParse({
      fields: [
        { foo: true, bar: "this is a valid field" },
        { foo: false, bar: "short" },
      ],
    }).success
  );
  t.end();
});
```

I would expect that to pass the validation as `foo` is `false`.

**Reply by alexbrokethis** [reply]
You're right. I've got it working on my end for something similar but when the variable field  (`foo` in your case) is an `enum`.

Another way you could solve it (tested and working in your codesandbox) is using a discriminated union:

```ts
const fooBar = z.object({
  foo: z.literal(true),
  // How do I validate this only if foo is true?
  bar: z.string().min(10, { message: "Should be 10 characters" }),
});

const fooBarOptional = z.object({
  foo: z.literal(false)
});

const fooBarUnion = z.discriminatedUnion("foo", [fooBar, fooBarOptional]);
```

sandbox [here](https://codesandbox.io/p/sandbox/staging-silence-ojufdb?file=%2Fsrc%2Findex.ts&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A42%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A1%7D%5D)

**Reply by chrisboakes** [reply]
@alexbrokethis thanks for helping with this! That sandbox link didn't work for me - it just goes to my original one. I [updated mine](https://codesandbox.io/p/sandbox/staging-silence-ojufdb?welcome=true) with that example but both tests fail. Maybe I'm missing something?

**Reply by alexbrokethis** [reply]
Ah, you need to keep your original schema and place `fooBarUnion` into it. Your tests are failing because you no longer have the `fields` array in your schema.

I'm having problems with codesandbox (clearly) so here's the full code:

```ts
import test from "tape";
import { z } from "zod";

const fooBar = z.object({
  foo: z.literal(true),
  // How do I validate this only if foo is true?
  bar: z.string().min(10, { message: "Should be 10 characters" }),
});

const fooBarOptional = z.object({
  foo: z.literal(false),
});

const fooBarUnion = z.discriminatedUnion("foo", [fooBar, fooBarOptional]);

const schema = z.object({
  fields: z.array(fooBarUnion),
});

test("valid when both are true", (t) => {
  t.true(
    schema.safeParse({
      fields: [
        { foo: true, bar: "this is a valid field" },
        { foo: true, bar: "this is a valid field" },
      ],
    }).success
  );
  t.end();
});

test("should not validate bar when foo is false", (t) => {
  t.true(
    schema.safeParse({
      fields: [
        { foo: true, bar: "this is a valid field" },
        { foo: false, bar: "short" },
      ],
    }).success
  );
  t.end();
});

```

**Reply by chrisboakes** [reply]
Ah sorry, I misinterpreted! That's exactly what I was after. Thanks so much!

**Comment by chrisboakes** [comment, marked as answer]
Resolved by @alexbrokethis: https://github.com/colinhacks/zod/discussions/2207#discussioncomment-5345805 Thanks! Working version [here](https://codesandbox.io/p/sandbox/staging-silence-ojufdb?file=%2Fsrc%2Findex.ts&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A42%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A1%7D%5D).
